### PR TITLE
[Actions] Run `brew build-bottle-pr` on merges to master

### PR DESCRIPTION
**Pull request by issyl0**
- Currently, after merges between Homebrew/homebrew-core and Homebrew/linuxbrew-core, maintainers must build bottles for formulae that had merge conflicts that were in the bottle blocks.
- Thankfully, there's a command for this: [`brew build-bottle-pr`](https://github.com/Linuxbrew/homebrew-developer/blob/master/cmd/brew-build-bottle-pr.rb), but it still takes time to get the list of changed formulae and run the command to raise each PR.
- GitHub Actions, on merges to master, can run [`brew find-formulae-to-bottle`](https://github.com/Linuxbrew/homebrew-developer/blob/master/cmd/brew-find-formulae-to-bottle.rb) to detect if it's a merge commit, and if it is, it will run `brew build-bottle-pr` on each of the changed formulae and raise PRs as the `LinuxbrewTestBot` GitHub user.

TODO:
- [ ] Generate GITHUB_TOKEN and SSH_KEY credentials for the LinuxbrewTestBot user and add them to its account. This requires a GITHUB_TOKEN because of `hub`. I am not sure if the default Actions-provided one will work. It might, because it'll be raising PRs against the same repo the Action is running from (unlike in my testing repo).
- [ ] Add the credentials to the secrets repo settings pane so that Actions can retrieve them. This requires a repo admin, please.

Things to note:
- This Action does not handle deleting of branches. This will become a problem _if_ we choose to stick with the default I chose here of the bot pushing branches to its fork:
  - Someone will have to log in as that user and delete the closed PRs branches.
  - Or, we set this up to push to branches directly on this repo (`origin`), which means that maintainers can delete them, or we can use [GitHub's new "delete branches" feature](https://help.github.com/en/articles/managing-the-automatic-deletion-of-branches) (though that might work better if we _merged_ PRs rather than _closed_ them?).
  - A future Action may be to make something that deletes branches automatically. :pray:
- THIS IS AS YET ENTIRELY UNTESTED WITH THE NEW ACTIONS YAML SYNTAX AND THE LINUXBREWTESTBOT USER!

### CODE REVIEW

File: .github/workflows/raise_bottle_prs.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+          brew help # trigger vendored ruby installation
+
+      - name: Configure git user
+        run: |
+          git config --global user.name "LinuxbrewTestBot"
+          git config --global user.email "testbot@linuxbrew.sh"
+
+      - name: Set up SSH key
+        env:
+          ISSYL0_BOT_SSH_KEY: {{ secrets.ISSYL0_BOT_SSH_KEY }}
```
Comment (issyl0) [inline .github/workflows/raise_bottle_prs.yml]: Missed this one. :flushed:

**Comment by xu-cheng** [comment]
FYI, this won't work. For security reason, the secret tokens should never be exposed to formulae building process. This is applied even if the maintainers have merged the PR. The reason is that there is no way you can validate entire build process to not leak the secrets. Simply put, we cannot trust any formulae with the secret tokens.

This is also why brew currently uses two step building process: one to build the formula without any secrets and one uploads the bottles.

**Comment by jonchang** [comment]
I believe this is intended to just open pull requests to build bottles on Linux when a merge commit from homebrew-core occurs.

**Comment by xu-cheng** [comment]
Oh, thanks for the clarification. I mistake the PR as building bottle directly. 

Nevertheless, this still exposes secret tokens to the untrusted formulae. Noted that the formulae will be loaded by `brew build-bottle-pr`.

To enhance the security, I think it would be better to use a separated account other than @linuxbrewtestbot. In this way, the token cannot be used to write to https://github.com/linuxbrewtestbot/homebrew-core. Otherwise, it may compromise our bottle/merge infrastructure.

**Comment by issyl0** [comment]
Thanks for your feedback, @xu-cheng. Thanks for the clarification of this PR's intentions, @jonchang - I was mid-way into typing a longer response myself but you covered it in one sentence! :-)

> To enhance the security, I think it would be better to use a separated account other than @LinuxbrewTestBot.

I understand your concerns. Making this be a separate user sounds better. Any suggestions for what we should call it? I think this is the only automatable and consistently performed workflow we have that raises PRs for things.

**Comment by MikeMcQuaid** [comment]
> Nevertheless, this still exposes secret tokens to the untrusted formulae. Noted that the formulae will be loaded by `brew build-bottle-pr`.

Is there any way this can be done without loading the formulae?

**Comment by MikeMcQuaid** [comment]
> To enhance the security, I think it would be better to use a separated account other than @LinuxbrewTestBot. In this way, the token cannot be used to write to https://github.com/linuxbrewtestbot/homebrew-core. Otherwise, it may compromise our bottle/merge infrastructure.

Note it could use a separate token with restricted privileges rather than a separate user.

**Comment by reitermarkus** [comment]
From what I suggested on Slack:

Use `GITHUB_TOKEN`, which is scoped to this repo and (as far as I know) unique for each action run.

Then push the branch directly to `linuxbrew-core`, which also has the advantage that maintainers can delete that branch with a button click.

**Comment by reitermarkus** [comment]
Like this, but using `GITHUB_TOKEN`:

https://github.com/Homebrew/homebrew-cask/blob/master/.github/actions/sync/templates.rb#L20-L28

**Comment by xu-cheng** [comment]
> Note it could use a separate token with restricted privileges rather than a separate user.

Currently, brew build-bottle-pr works by pushing to the https://github.com/linuxbrewtestbot/homebrew-core. This means the same token can also be used to rewrite any bottle PR. So you have to use a different account.

**Comment by reitermarkus** [comment]
> This means the same token can also be used to rewrite any bottle PR.

As I understand it, this command is only used for formulae from `homebrew-core`, so for a loaded formula to rewrite any bottle PR, it would already have to be infected in `homebrew-core`.

Anyways, rather than using yet another bot account, the obvious solution here is to split the formula loading from the PR creation or somehow not require loading the formula at all.

**Comment by xu-cheng** [comment]
My understanding is that once most of jobs are done by the CI automatically, no one will pay much attention to review the changes. Further, I am more worry the token leaked accidentally rather than by malicious adversaries. 

So I think the best solution would be making this work without loading the formulae. This would give a stronger guarantee. And we can use @linuxbrewtestbot as it is.

**Comment by reitermarkus** [comment]
> I am more worry the token leaked accidentally rather than by malicious adversaries.

From https://github.com/Homebrew/linuxbrew-core/pull/14838#issuecomment-521700158:

> Use `GITHUB_TOKEN`, which is scoped to this repo and (as far as I know) unique for each action run.

I still have to verify, but I assume it is uniquely generated for each run since there is no way to regenerate the `GITHUB_TOKEN` manually.

**Comment by issyl0** [comment]
@reitermarkus suggested extracting the [bits of `brew build-bottle-pr` that load formulae to check things]() into `brew find-formulae-to-bottle`, then the thing that pushes commits and uses the token doesn’t have to load formulae. I’ll work on this tonight, then hopefully this will be OK.

**Comment by dawidd6** [comment]
We can have an action that would check the commit message for "closes" keyword and extract the PR number that this commit is closing and then delete its head branch.
I already tested this procedure on my own tap and it seems to work just fine:

https://github.com/dawidd6/homebrew-tap/blob/master/.github/workflows/cleaning.yml

in fact, I even made a bunch of actions for that.

**Comment by issyl0** [comment]
Given that #15766 exists so that we think of a better approach to merge-based bottling, I'm going to close this for now until a consensus is reached on that. Maybe we can find a better way than the `build-bottle-pr` step?

Thanks to everyone to suggested things and audited for security!

**Pull request closed without merging**